cyral/resource_cyral_sidecar_listener.go

```diff
@@ -0,0 +1,428 @@
+	d.SetId(marshalComposedID([]string{d.Get(SidecarIdKey).(string), c.ListenerId}, "/"))
+	return d.Set(ListenerIdKey, c.ListenerId)
+}
+
+func (data ReadSidecarListenersAPIResponse) WriteToSchema(d *schema.ResourceData) error {
+	if data.ListenerConfig != nil {
+		_ = d.Set(ListenerIdKey, data.ListenerConfig.ListenerId)
+		_ = d.Set(RepoTypesKey, data.ListenerConfig.RepoTypesAsInterface())
+		_ = d.Set(UnixListenerKey, data.ListenerConfig.UnixSettingsAsInterface())
+		_ = d.Set(TcpListenerKey, data.ListenerConfig.TcpSettingsAsInterface())
+		_ = d.Set(MultiplexedKey, data.ListenerConfig.Multiplexed)
+		_ = d.Set(S3SettingsKey, data.ListenerConfig.S3SettingsAsInterface())
+		_ = d.Set(DynamoDbSettingsKey, data.ListenerConfig.DynamoDbSettingsAsInterface())
+	}
+	return nil
+}
+func (l *SidecarListener) RepoTypesAsInterface() []interface{} {
+	if l.RepoTypes == nil {
+		return nil
+	}
+	result := make([]interface{}, len(l.RepoTypes))
+	for i, v := range l.RepoTypes {
+		result[i] = v
+	}
+	return result
+}
+func (l *SidecarListener) RepoTypesFromInterface(anInterface []interface{}) {
+	repoTypes := make([]string, len(anInterface))
+	for i, v := range anInterface {
+		repoTypes[i] = v.(string)
+	}
+	l.RepoTypes = repoTypes
+}
+func (l *SidecarListener) TcpSettingsAsInterface() []interface{} {
+	if l.TcpListener != nil {
+		result := []interface{}{
+			map[string]interface{}{
+				HostKey: l.TcpListener.Host,
+				PortKey: l.TcpListener.Port,
+			},
+		}
+		return result
+	}
+	return nil
+}
+func (l *SidecarListener) TcpSettingsFromInterface(anInterface []interface{}) {
+	if len(anInterface) == 0 || anInterface == nil {
+		return
+	}
+	l.TcpListener = &TcpListener{
+		Host: anInterface[0].(map[string]interface{})[HostKey].(string),
+		Port: anInterface[0].(map[string]interface{})[PortKey].(int),
+	}
+}
+func (l *SidecarListener) UnixSettingsAsInterface() []interface{} {
+	if l.UnixListener != nil {
+		result := []interface{}{
+			map[string]interface{}{
+				FileKey: l.UnixListener.File,
+			},
+		}
+		return result
+	}
+	return nil
+}
+func (l *SidecarListener) UnixSettingsFromInterface(anInterface []interface{}) {
+	if len(anInterface) == 0 || anInterface == nil {
+		return
+	}
+	l.UnixListener = &UnixListener{
+		File: anInterface[0].(map[string]interface{})[FileKey].(string),
+	}
+}
+func (l *SidecarListener) MysqlSettingsAsInterface() []interface{} {
+	if l.MysqlSettings == nil {
+		return nil
+	}
+	return []interface{}{map[string]interface{}{
+		DbVersionKey:    l.MysqlSettings.DbVersion,
+		CharacterSetKey: l.MysqlSettings.CharacterSet,
+	}}
+}
+func (l *SidecarListener) MysqlSettingsFromInterface(anInterface []interface{}) {
+	if len(anInterface) == 0 {
+		return
+	}
+	if anInterface != nil {
+		l.MysqlSettings = &MysqlSettings{
+			DbVersion:    anInterface[0].(map[string]interface{})[DbVersionKey].(string),
+			CharacterSet: anInterface[0].(map[string]interface{})[CharacterSetKey].(string),
+		}
+	}
```

Comment:
We can simplify this by using the following instead: (Also applies to the other fields)
```suggestion
	if len(anInterface) == 0 {
		return
	}
        l.MysqlSettings = &MysqlSettings{
	        DbVersion:    anInterface[0].(map[string]interface{})[DbVersionKey].(string),
	        CharacterSet: anInterface[0].(map[string]interface{})[CharacterSetKey].(string),
        }
```